### PR TITLE
Feature/asb 750 examination timetable

### PR DESCRIPTION
**Pull request by christopher-semmens**

### CODE REVIEW

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+const expectNotFound = (res) => {
+	expect(res.render).toHaveBeenCalledWith('error/not-found');
+	expect(res.status).toHaveBeenLastCalledWith(404);
+};
+
+describe('controllers/projects/examination-timetable/controller', () => {
+	let req;
+	let res;
+
+	describe('examination-timetable controllers', () => {
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:32]: better describe ?

Comment (jonathangoulding) [reply packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:32]: should be the func name ?

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+			beforeEach(() => {
+				req = {
+					...mockReq(),
+					params: {
+						case_ref: 'mock case ref'
+					},
+					session: {}
+				};
+				res = mockRes();
+				jest.useFakeTimers().setSystemTime(new Date('2023-01-02'));
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:46]: duplicate timer set

maybe put the date in a const called dateOverride ? or something better

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+			});
+			describe('When getting the examination table page', () => {
+				describe('and there are no issues', () => {
+					beforeEach(async () => {
+						getAppData.mockResolvedValue({
+							data: {
+								CaseReference: 'mock_case_ref',
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:53]: move these into fixtures ? 

api responses.

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+						case_ref: 'mock case ref'
+					},
+					session: {}
+				};
+				res = mockRes();
+				jest.useFakeTimers().setSystemTime(new Date('2023-01-02'));
+			});
+			it('should render a not found page for missing events', async () => {
+				const req = {};
+				req.session = {
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:257]: why not put in the req object ?

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/get-event.component.test.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+const {
+	getEvent
+} = require('../../../../../../../../src/controllers/projects/examination-timetable/utils/events/event/get-event');
+
+describe('controllers/projects/examination-timetable/utils/events/event/get-event', () => {
+	describe('#getEvent', () => {
+		describe('When getting the data for an event', () => {
+			let result;
+			const mockEvent = {
+				dateOfEvent: '2023-01-02',
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/get-event.component.test.js:10]: these dates should be const and tell us what they are. 

i/e - today, tomorrow, yesterday, leap year, 1 month ago etc. 

Will help when we come back to them

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/helpers/get-event-deadline-submission-state.test.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
+const {
+	isEventDeadlineSubmissionOpen,
+	isPastEventTypeDeadline
+} = require('../../../../../../../../../src/controllers/projects/examination-timetable/utils/events/event/helpers/get-event-deadline-submission-state');
+
+describe('controllers/projects/examination-timetable/utils/events/event/helpers/get-event-deadline-submission-state', () => {
+	const datePresent = '2023-01-02';
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/helpers/get-event-deadline-submission-state.test.js:7]: nice, see above comment to do this to.

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/utils/get-event-state.test.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+			const mockEvent = {};
+			describe('and the event is not a deadline or the event deadline has not started', () => {
+				let result;
+				beforeEach(() => {
+					isEventDeadlineSubmissionOpen.mockReturnValue(false);
+					isPastEventTypeDeadline.mockReturnValue(false);
+					result = getEventState(mockEvent);
+				});
+				it('should return null', () => {
+					expect(result).toEqual({ isSubmissionOpen: false, tag: null });
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/event/utils/get-event-state.test.js:30]: maybe could read - 'should submission open = false and a tag of null'

or something like that.

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/get-events.component.test.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			let result;
+			const mockAppData = {
+				CaseReference: 'mock case reference',
+				dateOfNonAcceptance: '2023-01-01'
+			};
+			beforeEach(async () => {
+				jest.useFakeTimers().setSystemTime(new Date('2023-01-02'));
+				fetchEvents.mockResolvedValue([
+					{
+						dateOfEvent: '2023-01-01',
```
Comment (jonathangoulding) [inline packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/utils/events/get-events.component.test.js:28]: magic dates need vars associated

File: packages/forms-web-app/src/controllers/projects/examination-timetable/utils/events/utils/fetch-events.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+const { getTimetables } = require('../../../../../../services/timetable.service');
+
+const fetchEvents = async (caseRef) => {
+	const { data } = await getTimetables(caseRef);
+
+	return data?.timetables || [];
```
Comment (jonathangoulding) [inline packages/forms-web-app/src/controllers/projects/examination-timetable/utils/events/utils/fetch-events.js:6]: the || [] should mean the ? is not needed ?

Comment (jonathangoulding) [reply packages/forms-web-app/src/controllers/projects/examination-timetable/utils/events/utils/fetch-events.js:6]: have you tested and empty array returned ?

Comment (christopher-semmens) [reply packages/forms-web-app/src/controllers/projects/examination-timetable/utils/events/utils/fetch-events.js:6]: @jonathangoulding I have added a test for this now. The api can return undefined which will break the function if the optional chaining is not there.

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js (hunk 4 above)
Comment (christopher-semmens) [reply packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:257]: Pre-existing test.

File: packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js (hunk 3 above)
Comment (christopher-semmens) [reply packages/forms-web-app/__tests__/unit/controllers/projects/examination-timetable/controller.compoent.test.js:53]: Pre-existing test.

**Review by jonathangoulding — APPROVED** [review]

**Review by jonathangoulding — APPROVED** [review]

**Pull request merged**